### PR TITLE
Fixing accessibility issues in shitlist

### PR DESCRIPTION
**Pull request by alex-page**
### WHY are these changes introduced?

The shitlist contained a bunch of real accessibility issues that have not been fixed.

Fixes https://github.com/Shopify/polaris-react/issues/1458

### WHAT is this pull request doing?

Fixed accessibility issues in `DropZone`, `Form`, `Modal`, `Section`, `Page`, `Tabs`, `TextField` and `TopBar`.

### CODE REVIEW

File: src/components/Modal/README.md [hunk 1]
```diff
@@ -346,8 +346,8 @@ class ModalExample extends React.Component {
           onClose={this.toggleModal}
           title="Get a shareable link"
           primaryAction={{
-            content: 'Close',
-            onAction: this.toggleModal,
+            content: 'Copy link',
```
Comment (ry5n) [inline src/components/Modal/README.md]: Why this change? I would expect a modal dialog to either have a confirmation action paired with cancel, or a simple close button like in the original example. I would expect any of these actions to close the modal, and I don’t feel like the copy link action fits that expectation.

File: src/components/Modal/components/CloseButton/CloseButton.tsx [hunk 2]
```diff
@@ -11,15 +13,20 @@ export interface Props {
...
+
+function CloseButton({title = true, onClick, polaris: {intl}}: CombinedProps) {
   const className = classNames(
     styles.CloseButton,
     !title && styles.withoutTitle,
   );
 
   return (
     <button onClick={onClick} className={className}>
+      <VisuallyHidden>{intl.translate('Polaris.Common.close')}</VisuallyHidden>
```
Comment (ry5n) [inline src/components/Modal/components/CloseButton/CloseButton.tsx]: Nothing wrong with `VisuallyHidden` here, but [our recommendation in the style guide](https://polaris.shopify.com/components/images-and-icons/icon) is to use the `Icon` component’s `accessibilityLabel` prop.

File: src/components/Modal/README.md (hunk 1 above)
Comment (ry5n) [inline src/components/Modal/README.md]: Can you say more about this change? I would expect a modal dialog to either have a confirmation action paired with cancel, or a simple close button like in the original example.

File: src/components/OptionList/OptionList.tsx [hunk 3]
```diff
@@ -147,7 +147,7 @@ export class OptionList extends React.Component<CombinedProps, State> {
               {titleMarkup}
               <ul
                 className={styles.Options}
-                id={this.id}
+                id={`${this.id}-${sectionIndex}`}
```
Comment (ry5n) [inline src/components/OptionList/OptionList.tsx]: 👍

File: src/components/TextField/README.md [hunk 4]
```diff
@@ -733,7 +733,11 @@ class SeparateValidationErrorExample extends React.Component {
             <InlineError message={errorMessage} fieldID={textFieldID} />
           </div>
         </Stack.Item>
-        <Button icon={DeleteMinor} />
+        <Button
+          icon={DeleteMinor}
+          accessibilityLabel="Remove item"
+          labelHidden
```
Comment (ry5n) [inline src/components/TextField/README.md]: Does `Button` have a `labelHidden` prop?

File: src/locales/en.json [hunk 5]
```diff
@@ -118,7 +122,8 @@
 
     "Page": {
       "Header": {
-        "iconWarningMessage": "The icon prop has been removed from Page. Upload an application icon in the Shopify Partners Dashboard 'App setup' section instead."
+        "iconWarningMessage": "The icon prop has been removed from Page. Upload an application icon in the Shopify Partners Dashboard 'App setup' section instead.",
+        "rollupButton": "Toggle roll up"
```
Comment (ry5n) [inline src/locales/en.json]: This opens a popover with page-level actions, and I think we can do better with the copy. The menu may contain sections, where there is a convention that one of them is “More actions” (see Orders and Discounts list pages). Maybe just “Actions”?

File: src/components/TextField/README.md (hunk 4 above)
Comment (alex-page) [reply src/components/TextField/README.md]: Nope my bad. This line should be removed.
```suggestion
```

File: src/locales/en.json (hunk 5 above)
Comment (alex-page) [reply src/locales/en.json]:
```suggestion
        "rollupButton": "Actions"
```

File: src/components/Modal/components/CloseButton/CloseButton.tsx (hunk 2 above)
Comment (alex-page) [reply src/components/Modal/components/CloseButton/CloseButton.tsx]: I have fixed this up

File: src/components/Form/Form.tsx [hunk 6]
```diff
@@ -49,12 +52,15 @@ export default class Form extends React.PureComponent<Props> {
...
       target,
+      polaris: {intl},
     } = this.props;
     const autoCompleteInputs = normalizeAutoComplete(autoComplete);
 
     const submitMarkup = implicitSubmit ? (
       <VisuallyHidden>
-        <button type="submit" aria-hidden="true" />
+        <button type="submit" aria-hidden="true">
+          {intl.translate('Polaris.Common.submit')}
```
Comment (ry5n) [inline src/components/Form/Form.tsx]: If the button is aria hidden, it won’t be seen by screen reader users anyway, right? Is this intended, or should we be removing the `aria-hidden`?

Comment (alex-page) [reply src/components/Form/Form.tsx]: Thats true, however buttons always need content inside of them. This is probably a false positive but I would prefer the button to have content instead of being empty.

Comment (ry5n) [reply src/components/Form/Form.tsx]: Although I’d expect tests to pass with empty content and `aria-hidden="true"`, I’m cool with this fix.

File: src/components/DropZone/DropZone.tsx [hunk 7]
```diff
@@ -304,6 +305,9 @@ export class DropZone extends React.Component<CombinedProps, State> {
         {dragErrorOverlay}
         <div className={styles.Container}>{children}</div>
         <VisuallyHidden>
+          <label htmlFor={id}>
```
Comment (dpersing) [inline src/components/DropZone/DropZone.tsx]: A couple of items for the drop zone label:

- We have the `label` prop to allow consumers to add a visible label to the drop zone. This prop adds a proper `<label>` with a `for`/`id` relationship for the visually hidden input for the component. However, when the default label is added, it creates two labels for the same component in the output. Only one will get read (typically the first one in the DOM).
- How would the visually hidden default label work if there are multiple drop zone components on the same page? I think the `label` prop is there as a way to prevent duplicate labels.

Here's what the output looks like right now for the example showing a consumer-provided label:

```
<label id="DropZone12Label" for="DropZone12" class="Label-Text_SY9aZ">Theme files</label>
...
<span class="VisuallyHidden-VisuallyHidden_36MaW">
<label for="DropZone12">Upload file</label>
<input id="DropZone12" type="file" multiple="" autocomplete="off">
</span>
```

I'm thinking we could:

- Provide a default visually hidden label unless the consumer provides their own through the prop, and document recommending providing visual labels for drop zones, especially if there are multiple ones on the page
- Or make the `label` prop required, and allow consumers to make it visible or visually hidden

This is complicated by how the control had an additional visible button, and the input is off-screen when it gets focus, which is an issue (#762).

File: src/components/Modal/README.md [hunk 8]
```diff
@@ -364,17 +364,17 @@ class ModalExample extends React.Component {
               <Stack.Item fill>
                 <TextField
                   ref={this.bindNode}
-                  label=""
+                  label="Discount link"
```
Comment (dpersing) [inline src/components/Modal/README.md]: This is great, but I'm going to log a separate issue I've now noticed with how text field adds `<label>`, `aria-label`, and `aria-labelledby`, which are all duplicative.

File: src/components/Modal/components/CloseButton/CloseButton.tsx [hunk 9]
```diff
@@ -11,15 +12,23 @@ export interface Props {
...
+type CombinedProps = Props & WithAppProviderProps;
+
+function CloseButton({title = true, onClick, polaris: {intl}}: CombinedProps) {
   const className = classNames(
     styles.CloseButton,
     !title && styles.withoutTitle,
   );
 
   return (
-    <button onClick={onClick} className={className}>
```
Comment (dpersing) [inline src/components/Modal/components/CloseButton/CloseButton.tsx]: Looks like this is being addressed in PR #1582 (which I happened to get to first today).

File: src/components/Sheet/README.md [hunk 10]
```diff
@@ -207,9 +207,11 @@ class SheetExample extends React.Component {
                   </div>
                   <Scrollable style={{padding: '1.6rem', height: '100%'}}>
                     <ChoiceList
+                      title="Select a sales channel"
```
Comment (dpersing) [inline src/components/Sheet/README.md]: This is good. Making a note that it also caused me to find an unnecessary and invalid `aria-describedby` attribute on the `<fieldset>` output. Will write that up.

File: src/components/DropZone/DropZone.tsx (hunk 7 above)
Comment (alex-page) [reply src/components/DropZone/DropZone.tsx]: My bad this should be a quick fix!

Comment (alex-page) [reply src/components/DropZone/DropZone.tsx]: I have created a new issue for making it required and fixed the issue for now when there is no label provided it uses a sensible default.

File: src/components/Modal/README.md (hunk 8 above)
Comment (dpersing) [reply src/components/Modal/README.md]: New issue: #1587

File: src/components/Sheet/README.md (hunk 10 above)
Comment (dpersing) [reply src/components/Sheet/README.md]: New issue: #1589

File: src/components/Modal/components/CloseButton/CloseButton.tsx (hunk 9 above)
Comment (robin-drexler) [reply src/components/Modal/components/CloseButton/CloseButton.tsx]: Would it make sense to close https://github.com/Shopify/polaris-react/pull/1582 if the same thing is already done here? :)

File: UNRELEASED.md [hunk 11]
```diff
@@ -18,6 +18,7 @@ Use [the changelog guidelines](https://git.io/polaris-changelog-guidelines) to f
 
 ### Bug fixes
 
+- Fixed accessibility issues in `DropZone`, `Form`, `Modal`, `Section`, `Page`, `Tabs`, `TextField` and `TopBar` ([#1565](pulls/1565)).
```
Comment (kaelig) [inline UNRELEASED.md]:
```suggestion
- Fixed accessibility issues in `DropZone`, `Form`, `Modal`, `Section`, `Page`, `Tabs`, `TextField` and `TopBar` ([#1565](https://github.com/Shopify/polaris-react/pull/1565)).
```

File: src/components/Modal/components/CloseButton/CloseButton.tsx (hunk 9 above)
Comment (alex-page) [reply src/components/Modal/components/CloseButton/CloseButton.tsx]: @robin-drexler i'll merge your PR into here :)

File: src/components/Tabs/Tabs.tsx [hunk 12]
```diff
@@ -35,7 +38,7 @@ export interface State {
   tabToFocus: number;
 }
 
-export default class Tabs extends React.PureComponent<Props, State> {
+export class Tabs extends React.PureComponent<CombinedProps, State> {
```
Comment (AndrewMusgrave) [inline src/components/Tabs/Tabs.tsx]: We had to export classes in a previous TS version however I don't think that's the case anymore. We should be able to 🔥 it
```suggestion
class Tabs extends React.PureComponent<CombinedProps, State> {
```

File: src/components/Form/Form.tsx [hunk 13]
```diff
@@ -36,7 +37,9 @@ export interface Props {
   onSubmit(event: React.FormEvent<HTMLFormElement>): void;
 }
 
-export default class Form extends React.PureComponent<Props> {
+type CombinedProps = Props & WithAppProviderProps;
+
+export class Form extends React.PureComponent<CombinedProps, never> {
```
Comment (AndrewMusgrave) [inline src/components/Form/Form.tsx]: ^

```suggestion
class Form extends React.PureComponent<CombinedProps, never> {
```

File: src/components/DropZone/tests/DropZone.test.tsx [hunk 14]
```diff
@@ -80,47 +77,41 @@ describe('<DropZone />', () => {
...
-  it('calls the onDragEnter callback when a dragEnter event is fired', () => {
+  it('calls the onDragEnter callback when a dragenter event is fired', () => {
     const dropZone = mountWithAppProvider(<DropZone onDragEnter={spy} />);
-    const event = createEvent('dragenter', files);
-    dropZone.getDOMNode().dispatchEvent(event);
+    fireEvent({element: dropZone, eventType: 'dragenter'});
     expect(spy).toHaveBeenCalled();
   });
 
   it('calls the onDragOver callback when a dragOver event is fired', () => {
```
Comment (AndrewMusgrave) [inline src/components/DropZone/tests/DropZone.test.tsx]: If we're fixing the typo above as might as well do it here as well 😄 

```suggestion
  it('calls the onDragOver callback when a dragover event is fired', () => {
```

File: src/components/DropZone/tests/DropZone.test.tsx [hunk 15]
```diff
@@ -80,47 +77,41 @@ describe('<DropZone />', () => {
...
     const dropZone = mountWithAppProvider(<DropZone onDragOver={spy} />);
-    const event = createEvent('dragover', files);
-    dropZone.getDOMNode().dispatchEvent(event);
+    fireEvent({element: dropZone, eventType: 'dragover'});
     expect(spy).toHaveBeenCalled();
   });
 
   it('calls the onDragLeave callback when a dragLeave event is fired', () => {
```
Comment (AndrewMusgrave) [inline src/components/DropZone/tests/DropZone.test.tsx]: ^

```suggestion
  it('calls the onDragLeave callback when a dragleave event is fired', () => {
```

File: src/components/Navigation/components/Section/Section.tsx [hunk 16]
```diff
@@ -112,7 +112,6 @@ export default class Section extends React.Component<Props, State> {
             <span className={styles.Icon}>
               <Icon source={HorizontalDotsMinor} />
             </span>
-            {ariaLabel}
```
Comment (AndrewMusgrave) [inline src/components/Navigation/components/Section/Section.tsx]: Just for context, why was the label removed here?

File: src/components/DropZone/DropZone.tsx [hunk 17]
```diff
@@ -292,38 +293,32 @@ export class DropZone extends React.Component<CombinedProps, State> {
...
-      dropZoneMarkup
-    );
-
     return (
-      <Provider value={this.getContext}>{labelledDropzoneMarkup}</Provider>
+      <Provider value={this.getContext}>
+        <Labelled
+          id={id}
+          label={
+            label ? label : intl.translate('Polaris.DropZone.FileUpload.label')
```
Comment (AndrewMusgrave) [inline src/components/DropZone/DropZone.tsx]: As a general pattern, we tend to use variables rather than having the logic inline to keep the return clean

File: src/components/DropZone/DropZone.tsx [hunk 18]
```diff
@@ -292,38 +293,32 @@ export class DropZone extends React.Component<CombinedProps, State> {
...
+          }
+          action={labelAction}
+          labelHidden={label ? labelHidden : true}
```
Comment (AndrewMusgrave) [inline src/components/DropZone/DropZone.tsx]: ^

File: src/components/Navigation/components/Section/Section.tsx (hunk 16 above)
Comment (alex-page) [reply src/components/Navigation/components/Section/Section.tsx]: Duplicate label on the button container above.
